src/common-util/src/main/java/ee/ria/xroad/common/util/IsSubjectClientIdDecoder.java

```diff
@@ -0,0 +1,156 @@
+                return CertUtils.getSubjectClientId(cert);
+            }
+            return parseClientIdFromLegacyName(x500name);
+        }
+        return parseClientId(x500name);
+    }
+
+    /*
+     * The encoding for clientID: <ul> <li>C = FI (country code must be 'FI' when
+     * using this decoder)</li> <li>O = organization (must be present) <li>CN =
+     * memberCode (business code without "Y" prefix)</li> <li>serialNumber =
+     * instanceIdentifier;serverCode;memberClass </ul>
+     */
+
+    private static final Pattern SPLIT_PATTERN = Pattern.compile("=");
+
+    private static ClientId parseClientId(X500Name x500name) {
+        String c = getRDNValue(x500name, BCStyle.C);
+        if (!"IS".equals(c)) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain valid country code");
+        }
+
+        if (getRDNValue(x500name, BCStyle.O) == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain organization");
+        }
+
+        String memberCode = getRDNValue(x500name, BCStyle.CN);
+        if (memberCode == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain common name");
+        }
+        /* added 060718 JJB */
+        String ou = getRDNValue(x500name, BCStyle.OU);
+        if (ou == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE, "Certificate OU incorrect");
+        }
+
+        String serialNumber = getRDNValue(x500name, BCStyle.SERIALNUMBER);
+        if (serialNumber == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain serial number");
+        }
+
+        final String[] components = SPLIT_PATTERN.split(ou);
+        if (components.length != NUM_COMPONENTS) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name's attribute OU has invalid value");
+        }
+
+        // Note. components[1] = serverCode, unused
+        /*
+         * OU=instanceIdentifier=xroad-island serverCode=skra.is memberClass=island-gov
+         */
+        final int instanceIdString = 11;
+        final int memberClassString = 3;
+        return ClientId.create(components[1].substring(0, instanceIdString), // instanceId
+                components[memberClassString], // memberClass
+                memberCode);
+    }
+
+    /*
+     * The legacy encoding for clientID: <ul> <li>C = FI (country code must be 'FI'
+     * when using this decoder)</li> <li>O = instanceId</li> <li>OU =
+     * memberClass</li> <li>CN = memberCode (business code without "Y" prefix)</li>
+     * </ul>
+     */
+    private static ClientId parseClientIdFromLegacyName(X500Name x500name) {
```

Comment:
Finnish certificate encoding format has changed, and this function parses the legacy version. Is this really needed for IS certificates?

Reply:
I will look into this on monday morning.

Reply:
@jhyoty I have updated the code based on the KB article and I have removed the ```parseClientIdFromLegacyName``` function from the class. https://confluence.niis.org/pages/viewpage.action?pageId=4293076